### PR TITLE
Support uint16 and uint32 in to_torch

### PR DESCRIPTION
**Pull request by ctr-martemovTT**
### Ticket

https://github.com/tenstorrent/tt-metal/issues/31150

### Problem description

This PR was created to facilitate merging the following PR from the fork: https://github.com/tenstorrent/tt-metal/pull/32350

### What's changed

- Remove nanobind code that explicitly converted the tensor to the signed variant. This changes behavior for both torch and numpy results. 
- Update tests that relied on the torch/numpy tensor to have signed values as a result of the `to_torch` conversion. Use the `convert_to_signed_tensor` function to match the previous behavior. 

### Checklist

- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:mirror/MaximArtemovEPAM/maxim_artemov/TTNN-31150-uint16-uint32-to_torch)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:mirror/MaximArtemovEPAM/maxim_artemov/TTNN-31150-uint16-uint32-to_torch)
- [x] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:mirror/MaximArtemovEPAM/maxim_artemov/TTNN-31150-uint16-uint32-to_torch)
- [ ] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:mirror/MaximArtemovEPAM/maxim_artemov/TTNN-31150-uint16-uint32-to_torch)
  - [x] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:mirror/MaximArtemovEPAM/maxim_artemov/TTNN-31150-uint16-uint32-to_torch)
  - [ ] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:mirror/MaximArtemovEPAM/maxim_artemov/TTNN-31150-uint16-uint32-to_torch)
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: tests/ttnn/unit_tests/test_torch_conversion.py [hunk 1]
```diff
@@ -423,3 +423,99 @@ def test_from_numpy_conversion(device, shape, ttnn_dtype, numpy_dtype, ttnn_layo
...
+    [
+        ([np.iinfo(np.uint16).max], ttnn.uint16, torch.uint16),
+        ([np.iinfo(np.uint16).min], ttnn.uint16, torch.uint16),
+        ([np.iinfo(np.uint32).max], ttnn.uint32, torch.uint32),
+        ([np.iinfo(np.uint32).min], ttnn.uint32, torch.uint32),
+    ],
+)
+@pytest.mark.parametrize("ttnn_layout", [ttnn.TILE_LAYOUT, ttnn.ROW_MAJOR_LAYOUT])
+@pytest.mark.parametrize("with_device", [True, False])
+def test_torch_conversion_unsigned_edge_cases_fixed(
```
Comment (aliaksei-sala) [inline tests/ttnn/unit_tests/test_torch_conversion.py]: What does 'fixed' mean in a test names?

File: tests/ttnn/unit_tests/test_torch_conversion.py [hunk 2]
```diff
@@ -423,3 +423,99 @@ def test_from_numpy_conversion(device, shape, ttnn_dtype, numpy_dtype, ttnn_layo
...
+    device, tensor_data, ttnn_dtype, torch_input_type, ttnn_layout, with_device
+):
+    torch_input_tensor = torch.tensor(tensor_data, dtype=torch_input_type)
+    ttnn_input_tensor = ttnn.from_torch(
```
Comment (aliaksei-sala) [inline tests/ttnn/unit_tests/test_torch_conversion.py]: We can skip ttnn.from_torch by passing a Python list directly to the ttnn.Tensor constructor. 
This keeps the test scope cleaner and focused on what we're actually trying to validate.

File: tests/ttnn/utils_for_testing.py [hunk 3]
```diff
@@ -49,6 +52,33 @@ def construct_pcc_assert_message(message, expected_pytorch_result, actual_pytorc
...
+
+    else:
+        return tensor
+
+
+def _normalize_tensor(tensor):
+    if isinstance(tensor, ttnn.Tensor):
+        tensor = ttnn.to_torch(tensor)
+
+    if is_unsigned_tensor(tensor):
```
Comment (aliaksei-sala) [inline tests/ttnn/utils_for_testing.py]: We are performing the tensor type check twice: once in is_unsigned_tensor() and again in _post_to_torch_conversion(). The check in is_unsigned_tensor is redundant and could be removed.

File: tests/ttnn/utils_for_testing.py [hunk 4]
```diff
@@ -400,6 +436,19 @@ def maybe_trace(op_func, enable_trace, device):
     return output
 
 
+def is_unsigned_tensor(py_tensor):
+    return py_tensor.dtype in TORCH_INTEGER_DTYPES
+
+
+def match_type_post_conversion(roundtrip_tensor, py_tensor):
```
Comment (aliaksei-sala) [inline tests/ttnn/utils_for_testing.py]: The function match_type_post_conversion only requires py_tensor.dtype rather than the full tensor object. Additionally, 'post_conversion' is too specific. 
if the name reflect the function's actual logic, which is data type alignment/conversion for torch/np tensors, it will improve readability
Something like: `align_tensor_dtype(tensor, type)` or similar

**Review by bbradelTT — APPROVED** [review]
Please address @aliaksei-sala 's comments.

File: tests/ttnn/unit_tests/test_torch_conversion.py (hunk 1 above)
Comment (ctr-martemovTT) [reply tests/ttnn/unit_tests/test_torch_conversion.py]: **resolved**: removed the 'fixed' part, 'edge cases' is sufficiently self-explanatory.

File: tests/ttnn/unit_tests/test_torch_conversion.py (hunk 2 above)
Comment (ctr-martemovTT) [reply tests/ttnn/unit_tests/test_torch_conversion.py]: **resolved**: using the tensor constructor instead of the `from_torch`

File: tests/ttnn/utils_for_testing.py (hunk 3 above)
Comment (ctr-martemovTT) [reply tests/ttnn/utils_for_testing.py]: **resolved**: removed redundant check

File: tests/ttnn/utils_for_testing.py (hunk 4 above)
Comment (ctr-martemovTT) [reply tests/ttnn/utils_for_testing.py]: **resolved**: renamed the function and uses, made it so it accepts dtype instead of the tensor.

File: tests/ttnn/utils_for_testing.py [hunk 5]
```diff
@@ -71,6 +100,10 @@ def assert_with_pcc(expected_pytorch_result, actual_pytorch_result, pcc=0.9999):
     Raises:
         AssertionError: If the tensor shapes don't match or if the PCC is below the specified threshold
     """
+
+    expected_pytorch_result = _normalize_tensor(expected_pytorch_result)
```
Comment (aliaksei-sala) [inline tests/ttnn/utils_for_testing.py]: According to the function description, `assert_with_pcc` expects two Torch tensors.

Are there many places that violate this precondition? 
If not, we could update the tests to use the function correctly.

**Review by ayerofieiev-tt — APPROVED** [review]
